onnxruntime/core/graph/graph.cc

```diff
@@ -2355,31 +2355,37 @@ Status Graph::SetGraphInputsOutputs() {
 
         auto output_arg_iter = output_name_to_node_arg_index.find(input_arg->Name());
         if (output_name_to_node_arg_index.end() == output_arg_iter) {
-          // This input arg should be fed when running evaluation.
-          // it should be a graph input.
+          // This input arg is not the output of another node so must come from either a graph input or an initializer.
           const std::string& name = input_arg->Name();
+
           if (added_input_names.end() == added_input_names.find(name)) {
             // This graph input has not been added into <graph_inputs_>.
+            bool is_initializer = name_to_initial_tensor_.find(name) != name_to_initial_tensor_.end();
+
             if (!graph_inputs_manually_set_) {
-              graph_inputs_including_initializers_.push_back(input_arg);
+              // all initializers need matching inputs if IR version < 4.
+              // for IR version >= 4 any matching inputs for initializers need to be explicitly specified
+              if (!is_initializer || ir_version_ < 4) {
```

Comment:
Maybe it's better to not hardcode this here, but instead have a (single?) util that is related to this check

Reply:
Maybe having two helpers instead would be better. 
  CanOverrideInitializer
  InitializersRequireMatchingGraphInput

Problem is, the latter one would only be used here. 

It's somewhat cryptic to have !CanOverrideInitializer() in this condition instead of ir_version_ < 4.

---
In reply to: [299648708](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299648708) [](ancestors = 299648708)

Reply:
As this is the only place where the check is around adding a graph input for an initializers leaving it as is.